zebra-network/src/peer_set/set.rs

```diff
@@ -423,6 +423,13 @@ where
                     .ready_services
                     .get_index_mut(index)
                     .expect("preselected index must be valid");
+
+                // `Buffer`ed polls can fill up their Buffer reservations, causing hangs.
+                // See #1593 for details.
+                assert!(!format!("{:?}", type_of(&service)).contains("Buffer"),
+                    "Clients must not use tower::Buffer, because PeerSet uses 
+                     poll_ready multiple times before each call, which can cause hangs with buffers");
```

Comment:
The peer set type is static, so we only need to check it once.

Let's move it to the start of `PeerSet::new`, and do the checks on `Self`, so we don't have to use `type_of`:
```suggestion

                // Using `poll_ready` without `call` fills up `Buffer` reservations, causing hangs.
                // See #1593 for details.
                assert!(!type_name::<Self>().contains("Buffer"),
                    "Clients must not use tower::Buffer, because PeerSet uses 
                     `poll_ready` multiple times before each `call`, which causes buffer hangs");
```

Reply:
`Self` (title case) is the current type:
https://doc.rust-lang.org/std/keyword.Self.html

`self` (lower case) is the current object:
https://doc.rust-lang.org/std/keyword.self.html

Reply:
I dont think this will work. I added `dbg!(type_name::<Self>());` as the first line in `new()` and here is the type i am getting (it contains the "Buffer" string):

```
[zebra-network/src/peer_set/set.rs:126] type_name::<Self>() = "zebra_network::peer_set::set::PeerSet<tower::load::peak_ewma::PeakEwmaDiscover<futures_util::stream::stream::filter::Filter<futures_channel::mpsc::Receiver<core::result::Result<tower::discover::Change<std::net::addr::SocketAddr, zebra_network::peer::client::Client>, alloc::boxed::Box<dyn std::error::Error+core::marker::Sync+core::marker::Send>>>, futures_util::future::ready::Ready<bool>, zebra_network::peer_set::initialize::init<tower::load_shed::LoadShed<tower::buffer::service::Buffer<zebrad::components::inbound::Inbound, zebra_network::protocol::internal::request::Request>>>::{{closure}}::{{closure}}>>>"
```

Reply:
You're right, we don't need to check the outer `PeerSet` type, we need to check the inner `D::Service` type.

The previous code did by check using the `service` returned by:
```rust
                let (key, service) = self
                    .ready_services
                    .get_index_mut(index)
                    .expect("preselected index must be valid");
```

So we should be able to do:
```suggestion

                // Using `poll_ready` without `call` fills up `Buffer` reservations, causing hangs.
                // See #1593 for details.
                assert!(!type_name::<D::Service>().contains("Buffer"),
                    "Clients must not use tower::Buffer, because PeerSet uses `poll_ready` multiple times before each `call`, which causes buffer hangs");
```

Reply:
The type of `ready_services` is `IndexMap<D::Key, D::Service>`:
https://github.com/ZcashFoundation/zebra/blob/15698245e1184526482e1ccd38b8b2b2a3f26406/zebra-network/src/peer_set/set.rs#L89